### DISCUSSION TITLE
Migrate from 1024bit RSA .apk  - to -  2048bit RRSA .aab

### DISCUSSION THREAD

**Discussion by Mintonist** [airsdk/Adobe-Runtime-Support#1611] [General]
Hi!

Seems time to time this task becomes actual for some of AIR developers. 
I have an app since 2010 with a lot installs. It signed with 1024RSA key.

I am tying to migrate from 1024rsa .apk to 2048rsa .aab

I find @pol2095 experience [here](https://github.com/airsdk/Adobe-Runtime-Support/issues/934#issuecomment-874512524)

Please, who had already done it, answer these questions:

1. Before uploading .aab do I need to set "Use existing app signing key from Java KeyStore" in Play App Signing settings with previous 1024bit key? Or can I choose "No, Google Play don't manage my upload key" in support form (https://support.google.com/googleplay/android-developer/contact/key?hl=en)?
2. Do you have Google Play Games | Firebase integration? Do I need to update their settings with new SHA1 fingerprint? Do I need create new app in Firebase console or only add SHA1 fingerprint to existing?
3. What is change for players? Can they continue playing previous version (1024RSA) and install new (2048RSA) without problems? What happens with Google Play Games | Firebase integration for players with previous app version (1024RSA)?

Thx!

**Comment by marchbold** [comment]
Hi,

The way I did it is to setup Google Play to manage app signing, and upload the old 1024bit cert. 

Then I created a new 2048bit certificate to use as the upload certificate. 

See this guide for details:

https://airsdk.dev/docs/tutorials/platform/android/packaging-android-app-bundles#signing-and-certificates

Once done I started using the upload certificate for signing aab and local testing.

You don't have to change anything with firebase or other services, your users app will still get an app signed by Google Play with your 1024 bit certificate. 

However if you are testing your builds locally you will be using the upload certificate so the services will fail then. So it is worth adding the signature of the upload certificate to your firebase app configuration as well.

Cheers

**Reply by Mintonist** [reply]
Thx! Sounds interesting!
So no need to ask Google support? It is strange that google support doesn't suggest this approach to @pol2095
Are there any pitfalls?) No problems with work for any your ANEs?
Recently I have found something bad about different sign keys (in some issue discussions) but forget where(
It seems when I will do this in app console - it will be irreversibly.

**Reply by marchbold** [reply]
Yeah, but not sure what you are concerned about, just means Google will sign your apks they generate from your aab with you 1024 bit certificate. From the extensions perspective nothing has changed.

If you want to hold onto it and use an old version of Java to sign your apps locally you can still do this. But it's complicated as you need to have multiple versions of Java. 

And the upload certificate you can change at any point.

**Reply by Mintonist** [reply]
I find the comment - https://github.com/airsdk/Adobe-Runtime-Support/discussions/609#discussioncomment-286113
It is about fingerprint for API calls to Facebook. It is like as Play Games or Firebase configuration I mention before. 

Shortly: you can use 1024bit key as Play App Signing key and 2048bit key as Upload key (in App integrity config of Google play console). It solves all problems and no need to write to google support.

@[marchbold](https://github.com/marchbold) Can you add notice about case with 1024bit key to the https://airsdk.dev/docs/tutorials/platform/android/packaging-android-app-bundles#upload-certificate ?

**Reply by marchbold** [reply]
Yes Google specifies you can use a 1024bit or higher certificate for the app signing certificate but must use a 2048bit certificate for the upload certificate. I don't believe there is any difference with a 1024 certificate needed for the docs though? What exactly do you feel needs clarifying? 

Regarding the fingerprint, and to clarify here, several services use the certificate fingerprint or signature to verify your application and to ensure the requests are coming from the correct application etc. 

When you give your app certificate to Google they still use the same certificate to convert your aab into an apk that gets delivered to a user so the signature of your app doesn't change. Services like Facebook and Firebase will still see the same signature and nothing will change. 

However, if you start using a different upload certificate, you will need to take care when debugging locally as you may be signing your debug builds with the new upload certificate. In these cases your signature will be different and these services may fail. You can either:
- use the app certificate to sign and debug locally;
- or add the signature of the upload certificate to your services.

**Comment by a9848840** [comment]
@Mintonist 
I also have the same problem,

Is this method recommended for upgrading keys?
https://support.google.com/googleplay/android-developer/contact/key?hl=en

After updating the key, it can be downloaded from 1024ras.apk
through  google play update 2048bit RSA .aab?

The purpose of my update is because
android:targetSdkVersion="31"
I have to update 1024--->2048RSA-key
